replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/persistence/ReplicatorHistoryManagerImpl.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.impl.persistence;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.stream.Stream;
+import javax.ws.rs.NotFoundException;
+import org.codice.ditto.replication.api.Status;
+import org.codice.ditto.replication.api.data.ReplicationStatus;
+import org.codice.ditto.replication.api.impl.data.ReplicationStatusImpl;
+import org.codice.ditto.replication.api.persistence.ReplicatorHistoryManager;
+
+/** A metacard/catalog based persistence implementation of the ReplicatorHistoryManager interface */
+public class ReplicatorHistoryManagerImpl implements ReplicatorHistoryManager {
+
+  private final ReplicationPersistentStore persistentStore;
+
+  public ReplicatorHistoryManagerImpl(ReplicationPersistentStore persistentStore) {
+    this.persistentStore = persistentStore;
+  }
+
+  @Override
+  public ReplicationStatus create() {
+    return new ReplicationStatusImpl();
+  }
+
+  @Override
+  public ReplicationStatus get(String id) {
+    return persistentStore.get(ReplicationStatusImpl.class, id);
+  }
+
+  @Override
+  public Stream<ReplicationStatus> objects() {
+    return persistentStore.objects(ReplicationStatusImpl.class).map(ReplicationStatus.class::cast);
+  }
+
+  @Override
+  public ReplicationStatus getByReplicatorId(String id) {
+    return objects()
+        .filter(status -> status.getReplicatorId().equals(id))
+        .findFirst()
+        .orElseThrow(NotFoundException::new);
+  }
+
+  @Override
+  public void save(ReplicationStatus status) {
+    if (!(status instanceof ReplicationStatusImpl)) {
+      throw new IllegalArgumentException(
+          "Expected a ReplicationStatusImpl but got a " + status.getClass().getSimpleName());
+    } else {
+
+      try {
+        ReplicationStatus previous = getByReplicatorId(status.getReplicatorId());
+        ReplicationStatus updatedStatus = addStats(previous, status);
```

Comment:
❗️ This is going to be problematic as we are compounding stats info ourselves here and if multiple `saves` happens on multiple threads or in the future on multiple processes, we will be overwriting info saved by another and thus loosing stats. For cloud based development, you either want to save individual entries and sum them up when you want to display or get the database to help you by having it add or update data.

Keeping stats like this in a single row in the DB is never easy to do unless you can guarantee that only one is ever reading the previous and updating it before anybody can read it. In addition, it does require that every time you need to update, you are forced to load the previous version; thus making 2 DB calls. Easier to penalize the read than the write.